helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -335,15 +362,21 @@ private void addGeneratedMessageToMap(final Message message,
    */
   private void schedulePendingMessageCleanUp(
       final Map<String, Map<String, Message>> pendingMessagesToPurge, ExecutorService workerPool,
-      final HelixDataAccessor accessor) {
+      final HelixDataAccessor accessor, Map<String, Map<String, Message>> staleMessageMap) {
     workerPool.submit(new Callable<Object>() {
-      @Override public Object call() {
+      @Override
+      public Object call() {
         for (Map.Entry<String, Map<String, Message>> entry : pendingMessagesToPurge.entrySet()) {
           String instanceName = entry.getKey();
           for (Message msg : entry.getValue().values()) {
             if (accessor.removeProperty(msg.getKey(accessor.keyBuilder(), instanceName))) {
               LogUtil.logInfo(logger, _eventId, String
                   .format("Deleted message %s from instance %s", msg.getMsgId(), instanceName));
+              staleMessageMap.getOrDefault(msg.getTgtName(), Collections.emptyMap())
```

Comment:
So based on what I commented before, if we calculate from scratch in every pipeline, we won't need to maintain this map here, right? Or we will have some usage side this stage that we will need to update the content of the map at runtime?

Reply:
We do need for maintaining the history from previous pipelines. This could help better reduce stale messages.